apple/swift-markdown
# basic support for parsing Doxygen `\param` and `\returns` commands

Bug/issue #, if applicable: rdar://69835334

## Summary

This PR introduces a new parser mode, adding support for a minimal set of Doxygen commands, to facilitate transitioning from alternate documentation systems into systems that use Swift-Markdown, such as Swift-DocC.

The parsing is gated behind a new parse option, currently named `parseMinimalDoxygen` (name open for bikeshedding). When both this and `parseBlockDirectives` are given, the block-directive parser will start looking for a specific set of Doxygen commands. This PR adds support for two basic commands: `\param` and `\returns` (though the latter will also recognize `\return` and `\result`, as those are synonyms in Doxygen itself).

To facilitate testing, this PR also extends the `markdown-tool dump-tree` command with an `--experimental-parse-doxygen-commands` flag which turns on the behavior.

I also fixed two other issues in the block directive parser in the course of this PR:

- If a document begins with a code fence, the contents of that code fence were incorrectly scanned for block directives.
- If a line run in a document or block directive included a line that didn't start on the beginning of the line (e.g. when a directive is written on one line), the source location was incorrectly given as the beginning of the line.

## Dependencies

This requires another PR to integrate with Swift-DocC, which is yet to be written.

## Testing

As the new functionality is gated behind a new ParseOption, testing should also ensure that existing functionality is not changed without the new option set.

Steps:
1. Write a Markdown file with Doxygen command content in it. See the tests added in this PR for examples.
2. `swift run markdown-tool dump-tree --parse-block-directives --experimental-parse-doxygen-commands test.md`
3. Inspect the result to ensure that it looks valid.
4. Repeat with `--source-locations` and ensure that the source locations for command descriptions look valid.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.
